beacon_node/beacon_chain/src/graffiti_calculator.rs

```diff
@@ -0,0 +1,399 @@
+    ) -> Self {
+        Self {
+            beacon_graffiti,
+            execution_layer,
+            epoch_duration,
+            log,
+        }
+    }
+
+    /// Returns the appropriate graffiti to use for block production, prioritizing
+    /// sources in the following order:
+    /// 1. Graffiti specified by the validator client.
+    /// 2. Graffiti specified by the user via beacon node CLI options.
+    /// 3. The EL & CL client version string, applicable when the EL supports version specification.
+    /// 4. The default lighthouse version string, used if the EL lacks version specification support.
+    pub async fn get_graffiti(&self, validator_graffiti: Option<Graffiti>) -> Graffiti {
+        if let Some(graffiti) = validator_graffiti {
+            return graffiti;
+        }
+
+        match self.beacon_graffiti {
+            GraffitiOrigin::UserSpecified(graffiti) => graffiti,
+            GraffitiOrigin::Calculated(default_graffiti) => {
+                let Some(execution_layer) = self.execution_layer.as_ref() else {
+                    // Return default graffiti if there is no execution layer. This
+                    // shouldn't occur if we're actually producing blocks.
+                    crit!(self.log, "No execution layer available for graffiti calculation during block production!");
+                    return default_graffiti;
+                };
+
+                // The engine version cache refresh service ensures this will almost always retrieve this data from the
+                // cache instead of making a request to the execution engine. A cache miss would only occur if lighthouse
+                // has recently started or the EL recently went offline.
+                let engine_versions = match execution_layer
+                    .get_engine_version(Some(
+                        self.epoch_duration * ENGINE_VERSION_AGE_LIMIT_EPOCH_MULTIPLE,
+                    ))
+                    .await
+                {
+                    Ok(engine_versions) => engine_versions,
+                    Err(el_error) => {
+                        warn!(self.log, "Failed to determine execution engine version for graffiti"; "error" => ?el_error);
+                        return default_graffiti;
+                    }
+                };
+
+                let Some(engine_version) = engine_versions.first() else {
+                    // Got an empty array which indicates the EL doesn't support the method
+                    debug!(
+                        self.log,
+                        "Using default lighthouse graffiti: EL does not support {} method",
+                        ENGINE_GET_CLIENT_VERSION_V1;
+                    );
+                    return default_graffiti;
+                };
+                if engine_versions.len() != 1 {
+                    // More than one version implies lighthouse is connected to
+                    // an EL multiplexer. We don't support modifying the graffiti
```

Comment:
When you're connected to multiple ELs, you can't return a single version, it wouldn't make sense. You could hack together some kind of response to cram all the versions in, but that's a hack. Further, it's better for the CL to actually be able to detect if it's connected to multiple endpoints so that it can not do things like try to embed it in the graffiti when it doesn't make sense. But there are presumably other potential cases where the CL might want to have full information about what it's connected to (logging being one of them). By being able to include multiple, this method also serves as a general way to detect if lighthouse is connected to a multiplexer. No other method offers that.